crates/client/data-availability/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+pub mod avail;
+pub mod celestia;
+pub mod ethereum;
+mod sharp;
+pub mod utils;
+
+use std::collections::HashMap;
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+use anyhow::Result;
+use async_trait::async_trait;
+use ethers::types::{I256, U256};
+use futures::StreamExt;
+use sc_client_api::client::BlockchainEvents;
+use serde::Deserialize;
+use sp_api::ProvideRuntimeApi;
+use sp_runtime::traits::Block as BlockT;
+
+pub type StorageWrites<'a> = Vec<(&'a [u8], &'a [u8])>;
+
+pub struct DataAvailabilityWorker<B, C>(PhantomData<(B, C)>);
+
+#[derive(Debug, Copy, Clone, PartialEq, clap::ValueEnum)]
+pub enum DaLayer {
+    Celestia,
+    Ethereum,
+    Avail,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Deserialize, Default)]
+pub enum DaMode {
+    #[serde(rename = "validity")]
+    Validity,
+    #[serde(rename = "volition")]
+    Volition,
+    #[serde(rename = "validium")]
+    #[default]
+    Validium,
+}
+
+#[async_trait]
+pub trait DaClient: Send + Sync {
+    fn get_mode(&self) -> DaMode;
+    async fn last_published_state(&self) -> Result<I256>;
+    async fn publish_state_diff(&self, state_diff: Vec<U256>) -> Result<()>;
+}
+
+impl<B, C> DataAvailabilityWorker<B, C>
+where
+    B: BlockT,
+    C: ProvideRuntimeApi<B>,
+    C: BlockchainEvents<B> + 'static,
+{
+    pub async fn prove_current_block(da_mode: DaMode, client: Arc<C>, madara_backend: Arc<mc_db::Backend<B>>) {
+        let mut storage_event_st = client.storage_changes_notification_stream(None, None).unwrap();
```

Comment:
Why is it safe to unwrap?

Reply:
Can update to log this storage change notification failure, but we initialize the node with the intent to prove thee current block and then subsequently can't read from the storage change stream we are operating in a bad state.

Would be on the operator to check the logs and see that no proofs are being generated.

Reply:
Added an expect for the time being